File: src/neighborhoods.rs
use itertools::Itertools;
use crate::problem::{Problem, ProblemBox};

pub enum NeighborhoodType {
  Geometric,
  GeometricOverlap,
  Permutation
}

pub fn get_geometric_neighbors(problem_instance :&Problem) -> Vec<Problem> {
  let mut neighbors = Vec::new();

  // Iterate over all rectangles in all boxes
  for current_rect in problem_instance.boxes.iter().flat_map(|b| b.rectangles.iter()) {
    // Now iterate over all possible moves! A rect can be placed
    // ... in any box
    for (possible_box_idx, possible_box) in problem_instance.boxes.iter().enumerate() {
      // ... in any coordinate within this box
      for (x, y) in (0..possible_box.side_length).cartesian_product(0..possible_box.side_length) {
        // ... at any rotation
        for is_flipped in [true, false] {
          // Clone into a new neighbor
          let mut neighbor = problem_instance.clone();
          // Get the "current rect" in the new neighbor
          neighbor.move_rect(current_rect.get_id(), current_rect.get_box_idx(), x, y, possible_box_idx, is_flipped);
          neighbor.calculate_score();

          // TODO: skip infeasible neighbors for now
          if neighbor.score == 0 {
            continue;
          }

          neighbors.push(neighbor);
        }
      }
    }
  }
  
  return neighbors;
}

pub fn get_permutation_neighbors(problem_instance: &Problem) -> Vec<Problem> {
  // Idea: The solution is encoded in a long list of rectangles
  //       and we generate a valid solution by placing them top left to bottom right
  //       in the boxes. 
  let encoded_problem = encode_permutation_solution(problem_instance);

  // Compute *every* permutation
  let permutations = encoded_problem.iter().permutations(encoded_problem.len());

  // Decode them back to solutions
  let neighbors = permutations.map(|p| decode_permutation_solution(p));

  // Return
  return neighbors;
}

fn encode_permutation_solution(problem_instance :&Problem) -> Vec<ProblemBox> {
  // Make a long list of all problem rects
  let mut rects = Vec::new();
  for problem_box in problem_instance.boxes.iter() {
    for problem_rect in problem_box.rectangles.iter() {
      rects.push(problem_rect);
    }
  }
  // Sort them by box_id, x_coord, y_coord
  rects.sort_unstable_by_key(|r| (r.get_box_idx(), r.get_origin()));
  
  // Null the other values for these rects?

  // Return that list
  return rects;
}

fn decode_permutation_solution(boxes: Vec<ProblemBox>) -> Problem {
  // Initialize empty problem
  // Go through the rect list and place them in the lowest-possible box
  // Carry over other problem parameters?
  // Return new problem
}

File: propro/neighborhoods.py
'''
Contains different neighborhood types for the solution space.
'''

from enum import Enum, auto
from copy import deepcopy

from problem import BoxSolution, Rectangle, Box
from utils import flatten

class NeighborhoodDefinition(Enum):
  '''Enumeration type for all different neighborhood definitions'''
  GEOMETRIC = auto()
  GEOMETRIC_OVERLAP = auto()
  PERMUTATION = auto()

  def get_neighborhood_method(self):
    '''Returns the according neighborhood definition method for this type'''
    match self:
      case NeighborhoodDefinition.PERMUTATION: return get_permutation_neighbors
      case NeighborhoodDefinition.GEOMETRIC: return get_geometric_neighbors
      case NeighborhoodDefinition.GEOMETRIC_OVERLAP: return get_geometric_neighbors
      # TODO: for overlap, copy get_geometric_neighbors but get all,
      # not just free coordinates from a box? Incident edges don't make sense there too



def get_geometric_neighbors(solution: BoxSolution):
  '''
  Calculates neighbors of a solution by geometric means
  Moves every rectangle in every box to every possible coordinate
  '''
  neighbors = []

  # TODO: needs an option for a rect to be moved into a new box

  # Iterate over all rectangles in all boxes
  for current_box in sorted(solution.boxes.values(), key=lambda box: len(box.rects)):
    for current_rect in current_box.rects.values():
      # Now iterate over all possible moves! A rect can be placed
      # ... in any box
      for possible_box in solution.boxes.values():
        # ... in any free coordinate within this box
        for (x, y) in current_box.get_free_coordinates():
          # ... at any rotation
          for is_flipped in [True, False]:

            # Skip iteration if the rect would overflow to the right/bottom
            if x + current_rect.width > solution.side_length:
              continue
            if y + current_rect.height > solution.side_length:
              continue


            neighbor = deepcopy(solution) # <- spensy
            neighbor.move_rect(current_rect.id, current_box.id, x, y, possible_box.id, is_flipped)

            neighbors.append(neighbor)
            # Once a neighbor is found that reduces the main scoring criteria, early return?
            # Dramatically speeds up early convergence
            if neighbor.get_score()[0] < solution.get_score()[0]:
              print(f"Early returned with {len(neighbors)} neighbors")
              return neighbors
  print(f"Explored all {len(neighbors)} neighbors")
  return neighbors

def get_permutation_neighbors(solution: BoxSolution):
  '''
  Encodes the solution into a long list of rects that get placed from top left-to bottom-right
  in each box. Then computes permutations of this list and turns them back to solutions.
  '''
  encoded_rects = __encode_solution(solution)

  neighbors = []

  # Do every possible pairwise swap
  for i in range(len(encoded_rects) - 1):
    encoded_neighbor = deepcopy(encoded_rects)
    encoded_neighbor[i], encoded_neighbor[i + 1] = encoded_neighbor[i + 1], encoded_neighbor[i]
    neighbor = __decode_rect_list(encoded_neighbor, solution.side_length)
    neighbors.append(neighbor)

  # Also flip every possible rect
  for i in range(len(encoded_rects)):
    encoded_neighbor = deepcopy(encoded_rects)
    this_rect = encoded_neighbor[i]
    this_rect.width, this_rect.height = this_rect.height, this_rect.width
    neighbor = __decode_rect_list(encoded_neighbor, solution.side_length)
    neighbors.append(neighbor)

  return neighbors

def __encode_solution(solution: BoxSolution) -> list[Rectangle]:
  '''
  Turns the solution into a list of boxes
  '''
  return flatten([b.rects.values() for b in solution.boxes.values()])

def __decode_rect_list(rects: list[Rectangle], box_length: int) -> BoxSolution:
  '''
  Turns a list of rectangles into a valid solution to the box-rect problem.
  '''
  # Take rects one by one and put them into a new box..
  # Once one boundary is crossed, start with a new box
  boxes = [Box(0, box_length)]
  current_box = boxes[0]
  current_y = 0 # Current row's y index
  next_x = 0 # Next x coordinate for a box
  next_y = 0 # Next y index for a row
  # Go through rects until all have been processed
  for rect in rects:
    # Case 1: Rect fits into this row
    if next_x + rect.width < box_length and next_y + rect.height < box_length:
      # Update this rect's coordinates
      rect.x = next_x
      rect.y = current_y
      current_box.add_rect(rect)
      # Also update the next corodinate
      next_x += rect.width
      next_y = max(next_y, current_y + rect.height)
      continue
    #  Case 2: Rects overflows to the right, but fits into a next row within this box
    if next_x + rect.width >= box_length and next_y + rect.height < box_length:
      # NOTE: By specification this must fit here, box_length is guaranteed to be larger than any rect side
      rect.x = 0
      rect.y = next_y
      current_box.add_rect(rect)
      # Update coordinates for the next box
      next_x = rect.width
      current_y = next_y
      next_y = current_y + rect.height
      continue
    # Case 3: Rect does not fit into this box, create a new one and push it to boxes
    current_box = Box(len(boxes), box_length)
    rect.x = 0
    rect.y = 0
    current_box.add_rect(rect)
    boxes.append(current_box)
    # And set the next coordinates again
    next_x = rect.width
    current_y = 0
    next_y = rect.height

  # Construct box solution from list of boxes
  return BoxSolution(box_length, boxes)
